web/lib/tunnelv2.py

```diff
@@ -0,0 +1,176 @@
+                token_name, token_key = authorization.split()
+            except ValueError:
+                continue
+
+            if token_name.lower() == 'basic':
+                return authorization
+
+        return ''
+
+    @classmethod
+    def _get_user(cls, s_or_r: ScopeOrRequest) -> Optional[User]:
+        if isinstance(s_or_r, django.http.HttpRequest):
+            if s_or_r.user.is_authenticated:
+                return s_or_r.user
+
+        if 'user' in s_or_r:
+            if s_or_r['user'].is_authenticated:
+                return s_or_r['user']
+
+        return None
+
+    @classmethod
+    def get_printertunnel(
+        cls, s_or_r: ScopeOrRequest
+    ) -> PrinterTunnel:
+        subdomain_code = cls.get_subdomain_code(s_or_r)
+        port = cls.get_port(s_or_r)
+        auth_header = cls.get_authorization_header(s_or_r)
+
+        qs_kwargs = {
+            'printer__user__is_active': True,
+        }
+
+        if subdomain_code:
+            qs_kwargs['subdomain_code'] = subdomain_code
+        elif port:
+            qs_kwargs['port'] = port
+
+        qs = PrinterTunnel.objects.filter(
+            **qs_kwargs
+        ).select_related('printer', 'printer__user')
+
+        logging.debug(('get_printertunnel', port, subdomain_code, auth_header))
+
+        realm = (
+            f'tunnel {subdomain_code}'
+            if subdomain_code else
+            f'tunnel {port}'
+        )
+
+        try:
+            scheme, raw_token = auth_header.split()
+        except ValueError:
+            scheme, raw_token = None, None
+
+        if scheme and scheme.lower() == 'basic':
+            try:
+                username, password = base64.b64decode(
+                    raw_token).decode().split(':')
+            except (binascii.Error, ValueError):
+                raise TunnelAuthenticationError(
+                    'invalid token', realm=realm)
+
+            pt = qs.filter(
+                basicauth_username=username,
+                basicauth_password=password,
+                internal=False,
+            ).first()
+
+            if pt is None:
+                raise TunnelAuthenticationError(
+                    'invalid credentials', realm=realm)
+            return pt
+
+        user = cls._get_user(s_or_r)
+        if user is not None:
+            if user.is_authenticated:
+                pt = qs.filter(
+                    printer__user_id=user.id,
+                    internal=True,
+                ).first()
+
+                if pt is not None:
+                    return pt
+
+                # req is not using basic auth, no 401 error
+                raise django.http.Http404
+
+        # no basic auth, no session - force basic auth
+        # TODO: when we have proper integrations we might
+        # be able to convert this to a 404
+        raise TunnelAuthenticationError('missing credentials', realm=realm)
+
+    @classmethod
+    def is_tunnel_request(cls, s_or_r: ScopeOrRequest) -> bool:
```

Comment:
If we make TunnelV2 a separate app and a separate container, this is not necessary either.

Reply:
Let's do this later